Arrowgene.Ddon.Database/Sql/DdonMariaDb.cs

```diff
@@ -0,0 +1,149 @@
+                Logger.Error($"Failed to build connection string");
+                return false;
+            }
+
+            ReusableConnection = new MySqlConnection(_connectionString);
+
+            if (_settings.WipeOnStartup)
+            {
+                Logger.Info($"WipeOnStartup is currently not supported by '{_settings.Database}'.");
+            }
+
+            return true;
+        }
+
+        private string BuildConnectionString(DatabaseSetting settings)
+        {
+            MySqlConnectionStringBuilder builder = new MySqlConnectionStringBuilder
+            {
+                Server = settings.Host,
+                UserID = settings.User,
+                Password = settings.Password,
+                Database = settings.Database,
+                IgnoreCommandTransaction = true,
+                Pooling = true
+            };
+            string connectionString = builder.ToString();
+            Logger.Info($"Connection String: {connectionString}");
+            return connectionString;
+        }
+
+        protected override MySqlConnection OpenNewConnection()
+        {
+            MySqlConnection connection = new MySqlConnection(_connectionString);
+            connection.Open();
+            return connection;
+        }
+
+        protected override MySqlCommand Command(string query, MySqlConnection connection)
+        {
+            return new MySqlCommand(query, connection);
+        }
+
+        /// <summary>
+        /// Always returns the first generated ID in a multi-statement environment. Ideally should be used on a per-connection basis.
+        /// https://dev.mysql.com/doc/refman/8.0/en/information-functions.html#function_last-insert-id
+        /// </summary>
+        protected override long AutoIncrement(MySqlConnection connection, MySqlCommand command)
+        {
+            return command.LastInsertedId;
+        }
+
+        public override int Upsert(string table, string[] columns, object[] values, string whereColumn,
+            object whereValue,
+            out long autoIncrement)
+        {
+            throw new NotImplementedException();
+        }
+
+        protected override void AddParameter(MySqlCommand command, string name, ushort value)
+        {
+            AddParameter(command, name, value, DbType.UInt16);
+        }
+
+        protected override ushort GetUInt16(MySqlDataReader reader, string column)
+        {
+            return reader.GetUInt16(column);
+        }
+
+        protected override void AddParameter(MySqlCommand command, string name, uint value)
+        {
+            AddParameter(command, name, value, DbType.UInt32);
+        }
+
+        protected override uint GetUInt32(MySqlDataReader reader, string column)
+        {
+            return reader.GetUInt32(column);
+        }
+
+        protected override string SqlInsertOrIgnoreItem => $"INSERT IGNORE INTO \"ddon_item\" ({BuildQueryField(ItemFields)}) VALUES ({BuildQueryInsert(ItemFields)});";
```

Comment:
I feel like rather than overriding individual queries for each DBMS we should instead be using standard SQL. If there's no INSERT OR REPLACE outside of SQLite maybe we should do without it entirely and instead rewrite the code so regular INSERT or UPDATE statements are used

Reply:
Unfortunately it seems like the engines all use different incompatible syntax.
What we could do instead is do something slightly less efficient via subqueries and compound queries.

Subquery to replace IGNORE:
From:
`$"INSERT OR IGNORE INTO \"ddon_item\" ({BuildQueryField(ItemFields)}) VALUES ({BuildQueryInsert(ItemFields)});";`
To:
`$"INSERT INTO \"ddon_item\" ({BuildQueryField(ItemFields)}) SELECT {BuildQueryInsert(ItemFields)} WHERE NOT EXISTS (SELECT 1 FROM \"ddon_item\" WHERE \"uid\"=@uid);";`

Compound query to replace REPLACE/UPSERT => IGNORE + UPDATE:
From:

```
public bool ReplaceCharacterJobData(uint commonId, CDataCharacterJobData replacedCharacterJobData) {
            return ExecuteNonQuery(SqlReplaceCharacterJobData, command => {
                AddParameter(command, commonId, replacedCharacterJobData);
            }) == 1;
}
```

To:

```
public bool ReplaceCharacterJobData(uint commonId, CDataCharacterJobData replacedCharacterJobData){
            return InsertCharacterJobData(commonId, replacedCharacterJobData) || UpdateCharacterJobData(commonId, replacedCharacterJobData);
        }
        
public bool InsertCharacterJobData(uint commonId, CDataCharacterJobData insertedCharacterJobData){
            return ExecuteNonQuery(SqlInsertOrIgnoreCharacterJobData, command =>{
                AddParameter(command, commonId, insertedCharacterJobData);
            }) == 1;
        }
```

Reply:
Even if its less efficient, I think its the way to go if we want to support multiple DBMS. Otherwise we'd end up either maintaining three sets of sql operations manually or building our own ORM

Reply:
Done